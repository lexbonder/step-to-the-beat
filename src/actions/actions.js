export const saveAccessToken = token => ({
  type: 'SAVE_ACCESS_TOKEN',
  token
})

export const clearAccessToken = () => ({
  type: 'CLEAR_ACCESS_TOKEN'
})

export const logInUser = () => ({
  type: 'LOG_IN_USER',
  loggedIn: true
})

export const logOutUser = () => ({
  type: 'LOG_OUT_USER',
  loggedIn: false
})

export const saveUserName = userName => ({
  type: 'SAVE_USER_NAME',
  userName
})

export const selectSpm = spm => ({
  type: 'SELECT_SPM',
  spm
})

export const selectGenre = genre => ({
  type: 'SELECT_GENRE',
  genre
})

export const saveSpm = spm => ({
  type: 'SAVE_SPM',
  spm
})