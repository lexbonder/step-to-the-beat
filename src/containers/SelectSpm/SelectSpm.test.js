/* eslint-disable */
import React from 'react';
import { shallow } from 'enzyme';
import { SelectSpm, MSTP, MDTP } from './SelectSpm';
import { selectSpm, saveRecentSpm } from '../../actions/actions';

describe('SelectSpm', () => {
  let wrapper;
  let mockRecentSpms = [148, 160]
  let mockSaveRecentSpm = jest.fn()
  let mockSelectSpm = jest.fn()
  let mockHistory = { push: jest.fn() }

  beforeEach(() => {
    wrapper = shallow(<SelectSpm
      recentSpms={mockRecentSpms}
      saveRecentSpm={mockSaveRecentSpm}
      selectSpm={mockSelectSpm}
      history={mockHistory}
    />)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have default props', () => {
    const mockState = { selection: '' }
    expect(wrapper.state()).toEqual(mockState)
  })

  describe('componentDidUpdate', () => {
    it('should call getRecentSpms', () => {
      wrapper.instance().getRecentSpms = jest.fn()

      wrapper.instance().componentDidUpdate()

      expect(wrapper.instance().getRecentSpms).toHaveBeenCalled()
    })
  })

  describe('getRecentSpms', () => {
    it('should match the snapshot if there are recent SPMs', () => {
      expect(wrapper.instance().getRecentSpms()).toMatchSnapshot()
    })

    it('should match the snapshot if there are no SPMs', () => {
      const wrapper = shallow(<SelectSpm
        recentSpms={[]}
        saveRecentSpm={mockSaveRecentSpm}
        selectSpm={mockSelectSpm}
        history={mockHistory}
      />)
      expect(wrapper.instance().getRecentSpms()).toMatchSnapshot()
    })
  })

  describe('selectRecent', () => {
    it('should set selection with the value of event.target.innerText', () => {
      const mockEvent = {target: {innerText: '140'}}
      wrapper.instance().selectRecent(mockEvent)
      expect(wrapper.state().selection).toEqual(140)
    })
  })

  describe('handleClick', () => {
    it('should not call saveRecentSpm if the value in selection is already in this.props.recentSpms', () => {
      wrapper.state().selection = 148
      wrapper.instance().handleClick()
      expect(mockSaveRecentSpm).not.toHaveBeenCalled()
    })

    it('should call saveRecentSpm if the value in selection is not in this.props.recentSpms', () => {
      wrapper.state().selection = 140
      wrapper.instance().handleClick()
      expect(mockSaveRecentSpm).toHaveBeenCalledWith(140)
    })
    
    it('should call selectSpm with the value in this.state.selection', () => {
      wrapper.state().selection = 140
      wrapper.instance().handleClick()
      expect(mockSelectSpm).toHaveBeenCalledWith(140)
    })

    it('should redirect the user to /select-genre', () => {
      wrapper.instance().handleClick()
      expect(mockHistory.push).toHaveBeenCalledWith('/select-genre')
    })
  })

  describe('handleChange', () => {
    it('should parse the event value and store it in this.state.selection', () => {
      const mockEvent = {target: {value: '150'}}
      wrapper.instance().handleChange(mockEvent)
      expect(wrapper.state().selection).toEqual(150)
    })
  })

  describe('handleBackButton', () => {
    it('should redirect the user to /saved-playlists', () => {
      wrapper.instance().handleBackButton()
      expect(mockHistory.push).toHaveBeenCalledWith('/saved-playlists')
    })
  })

  describe('MSTP', () => {
    it('should return an object with the value it has in state', () => {
      const mockState = {recentSpms: mockRecentSpms}
      const mapped = MSTP(mockState)
      expect(mapped.recentSpms).toEqual(mockRecentSpms)
    })
  })

  describe('MDTP', () => {
    let mockDispatch = jest.fn()
    let result;

    beforeEach(() => {
      result = MDTP(mockDispatch)
    })

    it('should call dispatch when saveRecentSpm is called', () => {
      result.saveRecentSpm(150)
      expect(mockDispatch).toHaveBeenCalledWith(saveRecentSpm(150))
    })

    it('should call dispatch when selectSpm is called', () => {
      result.selectSpm(150)
      expect(mockDispatch).toHaveBeenCalledWith(selectSpm(150))
    })
  })
})