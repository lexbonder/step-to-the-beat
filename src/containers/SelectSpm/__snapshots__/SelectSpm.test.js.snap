// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectSpm getRecentSpms should match the snapshot if there are recent SPMs 1`] = `
Array [
  <li
    className="spm-list-item"
    onClick={[Function]}
>
    148
</li>,
  <li
    className="spm-list-item"
    onClick={[Function]}
>
    160
</li>,
]
`;

exports[`SelectSpm should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SelectSpm
    history={
        Object {
            "push": [Function],
          }
    }
    recentSpms={
        Array [
            148,
            160,
          ]
    }
    saveRecentSpm={[Function]}
    selectSpm={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2
          className="recent-spms"
>
          Your Recent SPMs
</h2>,
        <div
          className="ul-wrapper"
>
          <ul
                    className="spm-list"
          >
                    <li
                              className="spm-list-item"
                              onClick={[Function]}
                    >
                              148
                    </li>
                    <li
                              className="spm-list-item"
                              onClick={[Function]}
                    >
                              160
                    </li>
                    <Link
                              className="link-to-calc"
                              replace={false}
                              to="/calculations"
                    >
                              <li
                                        className="spm-list-item"
                              >
                                        Calculate a new SPM...
                              </li>
                    </Link>
          </ul>
</div>,
        <h3
          className="spm-instructions"
>
          Enter an SPM or select from above
</h3>,
        <input
          className="select-spm-input"
          name="selection"
          onChange={[Function]}
          placeholder="Enter SPM"
          type="number"
          value=""
/>,
        <div
          className="button-wrapper"
>
          <button
                    className="next-and-back-buttons"
                    onClick={[Function]}
          >
                    Cancel
          </button>
          <button
                    className="next-and-back-buttons"
                    disabled={true}
                    onClick={[Function]}
          >
                    Select a Genre
          </button>
</div>,
      ],
      "className": "SelectSpm",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Your Recent SPMs",
          "className": "recent-spms",
        },
        "ref": null,
        "rendered": "Your Recent SPMs",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ul
            className="spm-list"
>
            <li
                        className="spm-list-item"
                        onClick={[Function]}
            >
                        148
            </li>
            <li
                        className="spm-list-item"
                        onClick={[Function]}
            >
                        160
            </li>
            <Link
                        className="link-to-calc"
                        replace={false}
                        to="/calculations"
            >
                        <li
                                    className="spm-list-item"
                        >
                                    Calculate a new SPM...
                        </li>
            </Link>
</ul>,
          "className": "ul-wrapper",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              Array [
                <li
                  className="spm-list-item"
                  onClick={[Function]}
>
                  148
</li>,
                <li
                  className="spm-list-item"
                  onClick={[Function]}
>
                  160
</li>,
              ],
              <Link
                className="link-to-calc"
                replace={false}
                to="/calculations"
>
                <li
                                className="spm-list-item"
                >
                                Calculate a new SPM...
                </li>
</Link>,
            ],
            "className": "spm-list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": 148,
                "className": "spm-list-item",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": 148,
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": 160,
                "className": "spm-list-item",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": 160,
              "type": "li",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <li
                  className="spm-list-item"
>
                  Calculate a new SPM...
</li>,
                "className": "link-to-calc",
                "replace": false,
                "to": "/calculations",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Calculate a new SPM...",
                  "className": "spm-list-item",
                },
                "ref": null,
                "rendered": "Calculate a new SPM...",
                "type": "li",
              },
              "type": [Function],
            },
          ],
          "type": "ul",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Enter an SPM or select from above",
          "className": "spm-instructions",
        },
        "ref": null,
        "rendered": "Enter an SPM or select from above",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "select-spm-input",
          "name": "selection",
          "onChange": [Function],
          "placeholder": "Enter SPM",
          "type": "number",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="next-and-back-buttons"
              onClick={[Function]}
>
              Cancel
</button>,
            <button
              className="next-and-back-buttons"
              disabled={true}
              onClick={[Function]}
>
              Select a Genre
</button>,
          ],
          "className": "button-wrapper",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Cancel",
              "className": "next-and-back-buttons",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Cancel",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Select a Genre",
              "className": "next-and-back-buttons",
              "disabled": true,
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Select a Genre",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2
            className="recent-spms"
>
            Your Recent SPMs
</h2>,
          <div
            className="ul-wrapper"
>
            <ul
                        className="spm-list"
            >
                        <li
                                    className="spm-list-item"
                                    onClick={[Function]}
                        >
                                    148
                        </li>
                        <li
                                    className="spm-list-item"
                                    onClick={[Function]}
                        >
                                    160
                        </li>
                        <Link
                                    className="link-to-calc"
                                    replace={false}
                                    to="/calculations"
                        >
                                    <li
                                                className="spm-list-item"
                                    >
                                                Calculate a new SPM...
                                    </li>
                        </Link>
            </ul>
</div>,
          <h3
            className="spm-instructions"
>
            Enter an SPM or select from above
</h3>,
          <input
            className="select-spm-input"
            name="selection"
            onChange={[Function]}
            placeholder="Enter SPM"
            type="number"
            value=""
/>,
          <div
            className="button-wrapper"
>
            <button
                        className="next-and-back-buttons"
                        onClick={[Function]}
            >
                        Cancel
            </button>
            <button
                        className="next-and-back-buttons"
                        disabled={true}
                        onClick={[Function]}
            >
                        Select a Genre
            </button>
</div>,
        ],
        "className": "SelectSpm",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Your Recent SPMs",
            "className": "recent-spms",
          },
          "ref": null,
          "rendered": "Your Recent SPMs",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ul
              className="spm-list"
>
              <li
                            className="spm-list-item"
                            onClick={[Function]}
              >
                            148
              </li>
              <li
                            className="spm-list-item"
                            onClick={[Function]}
              >
                            160
              </li>
              <Link
                            className="link-to-calc"
                            replace={false}
                            to="/calculations"
              >
                            <li
                                          className="spm-list-item"
                            >
                                          Calculate a new SPM...
                            </li>
              </Link>
</ul>,
            "className": "ul-wrapper",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                Array [
                  <li
                    className="spm-list-item"
                    onClick={[Function]}
>
                    148
</li>,
                  <li
                    className="spm-list-item"
                    onClick={[Function]}
>
                    160
</li>,
                ],
                <Link
                  className="link-to-calc"
                  replace={false}
                  to="/calculations"
>
                  <li
                                    className="spm-list-item"
                  >
                                    Calculate a new SPM...
                  </li>
</Link>,
              ],
              "className": "spm-list",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "0",
                "nodeType": "host",
                "props": Object {
                  "children": 148,
                  "className": "spm-list-item",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": 148,
                "type": "li",
              },
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "host",
                "props": Object {
                  "children": 160,
                  "className": "spm-list-item",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": 160,
                "type": "li",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <li
                    className="spm-list-item"
>
                    Calculate a new SPM...
</li>,
                  "className": "link-to-calc",
                  "replace": false,
                  "to": "/calculations",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Calculate a new SPM...",
                    "className": "spm-list-item",
                  },
                  "ref": null,
                  "rendered": "Calculate a new SPM...",
                  "type": "li",
                },
                "type": [Function],
              },
            ],
            "type": "ul",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Enter an SPM or select from above",
            "className": "spm-instructions",
          },
          "ref": null,
          "rendered": "Enter an SPM or select from above",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "select-spm-input",
            "name": "selection",
            "onChange": [Function],
            "placeholder": "Enter SPM",
            "type": "number",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="next-and-back-buttons"
                onClick={[Function]}
>
                Cancel
</button>,
              <button
                className="next-and-back-buttons"
                disabled={true}
                onClick={[Function]}
>
                Select a Genre
</button>,
            ],
            "className": "button-wrapper",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Cancel",
                "className": "next-and-back-buttons",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Cancel",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Select a Genre",
                "className": "next-and-back-buttons",
                "disabled": true,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Select a Genre",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
